control-plane/subcommand/install-cni/command.go

```diff
@@ -0,0 +1,373 @@
+		return 1
+	}
+
+	// Get the dest file we will write to (the name can change)
+	destFileName, err := destConfigFile(srcFileName, c.logger)
+	if err != nil {
+		c.logger.Error("Unable get destination config file", "error", err)
+		return 1
+	}
+
+	// Get the correct mounted file paths from inside the container
+	srcFile := filepath.Join(install.MountedCNINetDir, srcFileName)
+	destFile := filepath.Join(install.MountedCNINetDir, destFileName)
+
+	// Append the consul configuration to the config that is there
+	err = appendCNIConfig(cfg, srcFile, destFile, c.logger)
+	if err != nil {
+		c.logger.Error("Unable add the consul-cni config to the config file", "error", err)
+		return 1
+	}
+
+	// Generate the kubeconfig file that will be used by the plugin to communicate with the kubernetes api
+	err = createKubeConfig(install.MountedCNINetDir, cfg.Kubeconfig, c.logger)
+	if err != nil {
+		c.logger.Error("Unable to create kubeconfig file", "error", err)
+		return 1
+	}
+
+	// copy the consul-cni binary from the installer container to the host
+	err = copyCNIBinary(install.CNIBinSourceDir, install.MountedCNIBinDir, c.logger)
+	if err != nil {
+		c.logger.Error("Unable to copy cni binary", "error", err)
+		return 1
+	}
+
+	return 0
+
+}
+
+// newCNIConfig creates the consul-cni configuration based on flags that are passed to the command.
+func (c *Command) newCNIConfig() (*config.CNIConfig, error) {
+	return &config.CNIConfig{
+		Name:       defaultName,
+		Type:       defaultType,
+		CNIBinDir:  c.flagCNIBinDir,
+		CNINetDir:  c.flagCNINetDir,
+		Multus:     c.flagMultus,
+		Kubeconfig: c.flagKubeconfig,
+		LogLevel:   c.flagLogLevel,
+	}, nil
+}
+
+// newInstallConfig creates the install configuration that is needed for host level operations.
+func (c *Command) newInstallConfig() (*installConfig, error) {
+	return &installConfig{
+		MountedCNIBinDir: "/host" + c.flagCNIBinDir,
+		MountedCNINetDir: "/host" + c.flagCNINetDir,
+		CNIBinSourceDir:  c.flagCNIBinSourceDir,
+	}, nil
+}
+
+// appendCNIConfig appends the consul-cni configuration to the main configuration file.
+func appendCNIConfig(cfg *config.CNIConfig, srcFile, destFile string, logger hclog.Logger) error {
+
+	// Needed to convert the config struct for inserting
+	// Check if file exists
+	if _, err := os.Stat(srcFile); os.IsNotExist(err) {
+		return fmt.Errorf("source cni config file %s does not exist: %v", srcFile, err)
+	}
+	logger.Debug("appendCNIConfig: using files", "srcFile", srcFile, "destFile", destFile)
+	// This section overwrites an existing plugins list entry for istio-cni
+	existingCNIConfig, err := os.ReadFile(srcFile)
+	if err != nil {
+		return err
+	}
+
+	// convert the consul-cni struct into a map
+	var cfgMap map[string]interface{}
+	err = mapstructure.Decode(cfg, &cfgMap)
+	if err != nil {
+		return fmt.Errorf("error loading Consul CNI config: %v", err)
+	}
+
+	// Convert the json config file into a map. The map that is created has 2 parts:
+	// [0] the cni header ()
```

Comment:
Yes, they are. I had a hard time getting it working correctly. It could probably use a refactor but it was all difficult to work with as the plugins are all free form.